src/restic/hardlinks_index.go

```diff
@@ -0,0 +1,57 @@
+package restic
+
+import (
+	"sync"
+)
+
+// HardlinkKey is a composed key for finding inodes on a specific device
+type HardlinkKey struct {
+	Inode, Device uint64
+}
+
+// HardlinkIndex contains a list of inodes, devices these inodes are one, and associated file names
+type HardlinkIndex struct {
+	m     sync.Mutex
+	Index map[HardlinkKey]string
+}
+
+// NewHardlinkIndex create a new index for hard links
+func NewHardlinkIndex() *HardlinkIndex {
+	return &HardlinkIndex{
+		Index: make(map[HardlinkKey]string),
+	}
+}
+
+// ExistsLink checks wether the link already exist in the index
+func (idx *HardlinkIndex) ExistsLink(inode uint64, device uint64) bool {
```

Comment:
I'd prefer this function to be named `Has()` instead of `ExistsLink()`, that's the same style as [the index](https://godoc.org/github.com/restic/restic/src/restic/repository#Index.Has).